libavcodec/mfdec.c

```diff
@@ -0,0 +1,1717 @@
+#include "libavutil/imgutils.h"
+#include "libavutil/opt.h"
+#include "libavutil/time.h"
+#include "libavutil/internal.h"
+#include "internal.h"
+#include "codec_internal.h"
+#include "compat/w32dlfcn.h"
+#include "libavutil/hwcontext.h"
+#include "libavutil/hwcontext_mediafoundation.h"
+#include "mpeg4audio.h"
+
+// Used to destroy the decoder once the last frame reference has been
+// released when using opaque decoding mode.
+typedef struct MFDecoder {
+    IMFTransform *mft;
+    AVBufferRef *device_ref;
+} MFDecoder;
+
+typedef struct MFContext {
+    BASE_CODEC_CONTEXT
+    int lavc_init_done;
+    uint8_t* send_extradata;
+    int send_extradata_size;
+    AVBSFContext* bsfc;
+    int sw_format;
+    int use_opaque; // whether AV_PIX_FMT_MF is returned to the user
+    AVBufferRef* device_ref; // really AVHWDeviceContext*
+    AVBufferRef* frames_ref; // really AVHWFramesContext*
+    AVBufferRef* decoder_ref; // really MFDecoder*
+    // Important parameters which might be overwritten by decoding.
+    int original_channels;
+    // set by AVOption
+    int opt_use_d3d;
+    int opt_require_d3d;
+    int opt_out_samples;
+    int opt_d3d_bind_flags;
+    AVFrame* tmp_frame;
+} MFContext;
+
+static int mf_choose_output_type(AVCodecContext *avctx);
+static int mf_setup_context(AVCodecContext *avctx);
+
+#define MF_TIMEBASE (AVRational){1, 10000000}
+// Sentinel value only used by us.
+#define MF_INVALID_TIME AV_NOPTS_VALUE
+
+static int mf_wait_events(AVCodecContext *avctx)
+{
+    MFContext *c = avctx->priv_data;
+
+    if (!c->async_events)
+        return 0;
+
+    while (!(c->async_need_input || c->async_have_output || c->draining_done || c->async_marker)) {
+        IMFMediaEvent *ev = NULL;
+        MediaEventType ev_id = 0;
+        HRESULT hr = IMFMediaEventGenerator_GetEvent(c->async_events, 0, &ev);
+        if (FAILED(hr)) {
+            av_log(avctx, AV_LOG_ERROR, "IMFMediaEventGenerator_GetEvent() failed: %s\n",
+                   ff_hr_str(hr));
+            return AVERROR_EXTERNAL;
+        }
+        IMFMediaEvent_GetType(ev, &ev_id);
+        switch (ev_id) {
+        case ff_METransformNeedInput:
+            if (!c->draining)
+                c->async_need_input = 1;
+            break;
+        case ff_METransformHaveOutput:
+            c->async_have_output = 1;
+            break;
+        case ff_METransformDrainComplete:
+            c->draining_done = 1;
+            break;
+        case ff_METransformMarker:
+            c->async_marker = 1;
+            break;
+        default: ;
+        }
+        IMFMediaEvent_Release(ev);
+    }
+
+    return 0;
+}
+
+static AVRational mf_get_tb(AVCodecContext *avctx)
+{
+    if (avctx->pkt_timebase.num > 0 && avctx->pkt_timebase.den > 0)
+        return avctx->pkt_timebase;
+    if (avctx->time_base.num > 0 && avctx->time_base.den > 0)
+        return avctx->time_base;
+    return MF_TIMEBASE;
+}
+
+static LONGLONG mf_to_mf_time(AVCodecContext *avctx, int64_t av_pts)
```

Comment:
I will take a look here, this might have been a holdover from the original patch from WM4.  Regardless, it should be readable and make sense as you suggest.